### PR TITLE
Accessibility configuration Added for Exclude Rule from analysis

### PR DESCRIPTION
**Pull request by prashelke**
Thank you for your contribution.
Before submitting this PR, please make sure:

- [x] PR description and commit message should describe the changes done in this PR
- [x] Verify the PR is pointing to correct branch i.e. Release or Beta branch if the code fix is for specific release , else point it to master
- [x] Latest Code from master or specific release branch is merged to your branch
- [x] No unwanted\commented\junk code is included
- [x] No new warning upon build solution
- [x] Code Summary\Comments are added to my code which explains what my code is doing
- [ ] Existing unit test cases are passed
- [x] New Unit tests are added for your development
- [x] Sanity Tests are successfully executed for New and Existing Functionality
- [x] Verify that changes are compatible with all relevant browsers and platforms.
- [x] After creating pull request there should not be any conflicts
- [x] Resolve all Codacy comments
- [x] Builds and checks are passed before PR is sent for review
- [ ] Resolve code review comments
- [ ] Update the Help Library document to match any feature changes

## Summary by CodeRabbit

- **New Features**
  - Introduced an accessibility rules menu item for improved navigation.
  - Added a new user interface page for managing accessibility rules.
  - Implemented functionality for dynamic management of accessibility configurations.

- **Improvements**
  - Enhanced the handling of accessibility rules and configurations for better clarity and maintainability.
  - Updated user interface elements for improved aesthetics and usability.
  - Streamlined accessibility testing processes and reporting functionalities.

- **Tests**
  - Created a comprehensive suite of unit tests to evaluate accessibility features and compliance.

### CODE REVIEW

File: Ginger/Ginger/MenusLib/ConfigurationsMenu.cs [hunk 1]
```diff
@@ -155,5 +159,10 @@ private static Page OthersPage()
         {
             return new Page();
         }
+
+        private static Page GeAccessiblityRulePage()
+        {
+            return (new AccessiblityRulePage());
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/MenusLib/ConfigurationsMenu.cs:165]: done

File: Ginger/GingerCoreCommon/External/Configurations/AccessibilityConfiguration.cs [hunk 2]
```diff
@@ -29,231 +29,25 @@ namespace Ginger.Configurations
...
-            {
-                return _SeverityItems;
-            }
-            set
-            {
-                _SeverityItems = value;
-
-            }
-        }
+        public override string ItemName { get => throw new System.NotImplementedException(); set => throw new System.NotImplementedException(); }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreCommon/External/Configurations/AccessibilityConfiguration.cs:32]: done

File: Ginger/Ginger/SolutionWindows/AccessiblityRulePage.xaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+﻿<Page x:Class="Ginger.SolutionWindows.AccessiblityRulePage"
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/SolutionWindows/AccessiblityRulePage.xaml:1]: done

File: Ginger/GingerCoreCommon/External/Configurations/AccessibilityRuleData.cs [hunk 4]
```diff
@@ -62,10 +63,22 @@ public class AccessibilityRuleData : RepositoryItemBase, INotifyPropertyChanged
         public string ACTRules { get { return mACTRules; } set { if (mACTRules != value) { mACTRules = value; OnPropertyChanged(nameof(ACTRules)); } } }
         public override string ItemName { get; set; }
 
+        public ObservableList<AccessibilityRuleData> GetAccessibilityRules(string AccessbiltyString)
+        {
+            ObservableList<AccessibilityRuleData> accessibilityRules = new ObservableList<AccessibilityRuleData>();
+            Root data = JsonConvert.DeserializeObject<Root>(AccessbiltyString);
+            return data.accessibilityRules;
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreCommon/External/Configurations/AccessibilityRuleData.cs:71]: done

File: Ginger/Ginger/SolutionWindows/AccessiblityRulePage.xaml.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        {
+            InitializeComponent();
+            mSolution = WorkSpace.Instance.Solution;
+            string allProperties = string.Empty;
+            WorkSpace.Instance.PropertyChanged += WorkSpacePropertyChanged;
+            LoadGridData();
+            SetAppsGrid();
+            mAccessibilityConfiguration = new();
+            DefaultExcludeRulesList = WorkSpace.Instance.Solution.DefaultExcludeRule.DefaultExcludeRules != null ? WorkSpace.Instance.Solution.DefaultExcludeRule.DefaultExcludeRules.Split(',').ToList() : new();
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/SolutionWindows/AccessiblityRulePage.xaml.cs:45]: done

File: Ginger/Ginger/SolutionWindows/AccessiblityRulePage.xaml.cs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+                {
+                    if (DefaultExcludeRulesList.Any(x => x.Equals(data.RuleID)))
+                    {
+                        DefaultExcludeRulesList.Remove(data.RuleID);
+                        mAccessibilityConfiguration.DefaultExcludeRules = String.Join(",", DefaultExcludeRulesList.Select(x => x));
+                        WorkSpace.Instance.Solution.DefaultExcludeRule = mAccessibilityConfiguration;
+                    }
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/SolutionWindows/AccessiblityRulePage.xaml.cs:140]: done

File: Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+            string allProperties = string.Empty;
+            WorkSpace.Instance.PropertyChanged += WorkSpacePropertyChanged;
+            LoadGridData();
+            SetAppsGrid();
+            mAccessibilityConfiguration = WorkSpace.Instance.SolutionRepository.GetFirstRepositoryItem<AccessibilityConfiguration>();
+            mAccessibilityConfiguration.DefaultExcludeRule = mAccessibilityConfiguration.DefaultExcludeRule != null ? mAccessibilityConfiguration.DefaultExcludeRule : new ();
+        }
+        private void WorkSpacePropertyChanged(object sender, System.ComponentModel.PropertyChangedEventArgs e)
+        {
+            if (e.PropertyName == nameof(WorkSpace.Solution))
```
Comment (Maheshkale447) [inline Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs:51]: As values are not being saved in solution xml, do we need this property change event?

File: Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+                AllowSorting = true
+            });
+            view.GridColsView.Add(new GridColView()
+            {
+                Field = nameof(AccessibilityRuleData.Tags),
+                WidthWeight = 30,
+                ReadOnly = true
+            });
+
+            xAccessibilityRulesGrid.AddLabel("Note: We will analyze the active items from the list below for accessibility. Testing and deactivating items won't be taken into account for accessibility testing.");
```
Comment (Maheshkale447) [inline Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs:101]: Change it to "Note: Ginger will only analyze the active accessibility testing rules; inactive rules will not be considered."

File: Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+            xAccessibilityRulesGrid.SetAllColumnsDefaultView(view);
+            xAccessibilityRulesGrid.InitViewItems();
+        }
+
+        private void LoadGridData()
+        {
+            ActAccessibilityTesting actAccessibilityTesting = new ActAccessibilityTesting();
```
Comment (Maheshkale447) [inline Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs:109]: Check if this class can be declared as static as there won't be multiple instance of this class

File: Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+            List<AccessibilityRuleData> sortedList = actAccessibilityTesting.RulesItemsdata.OrderByDescending(data => !data.Active).ToList();
```
Comment (Maheshkale447) [inline Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs:110]: Add method in ActAccessibilityTesting to get the active rules list

File: Ginger/GingerCoreCommon/External/Configurations/AccessibilityConfiguration.cs [hunk 11]
```diff
@@ -29,231 +30,42 @@ namespace Ginger.Configurations
...
+        
 
-            }
-        }
 
-        private ObservableList<OperationValues> mOperationValueList;
+        ObservableList<AccessibilityRuleData> mDefaultExcludeRule;
         [IsSerializedForLocalRepository]
-        public ObservableList<OperationValues> OperationValueList
+        public ObservableList<AccessibilityRuleData> DefaultExcludeRule
```
Comment (Maheshkale447) [inline Ginger/GingerCoreCommon/External/Configurations/AccessibilityConfiguration.cs:41]: Change its name to ExcludedRules

File: Ginger/GingerCoreCommon/External/Configurations/AccessibilityRuleData.cs [hunk 12]
```diff
@@ -62,10 +65,32 @@ public class AccessibilityRuleData : RepositoryItemBase, INotifyPropertyChanged
         public string ACTRules { get { return mACTRules; } set { if (mACTRules != value) { mACTRules = value; OnPropertyChanged(nameof(ACTRules)); } } }
         public override string ItemName { get; set; }
 
+        public ObservableList<AccessibilityRuleData> GetAccessibilityRules(string AccessbiltyString)
+        {
+            Root data = new();
```
Comment (Maheshkale447) [inline Ginger/GingerCoreCommon/External/Configurations/AccessibilityRuleData.cs:70]: change initialization to []

File: Ginger/GingerCoreCommon/Repository/GingerSolutionRepository.cs [hunk 13]
```diff
@@ -66,6 +67,8 @@ public static SolutionRepository CreateGingerSolutionRepository()
 
             SR.AddItemInfo<RemoteServiceGrid>("*.Ginger.RemoteServiceGrid.xml", SolutionRepository.cSolutionRootFolderSign + "RemoteServiceGrid", true, "RemoteServiceGrid", PropertyNameForFileName: nameof(RemoteServiceGrid.Name));
 
+            SR.AddItemInfo<AccessibilityConfiguration>("*.Ginger.Configuration.xml", SolutionRepository.cSolutionRootFolderSign + "Configurations", true, "AccessibilityConfiguration", PropertyNameForFileName: nameof(AccessibilityConfiguration.Name));
```
Comment (Maheshkale447) [inline Ginger/GingerCoreCommon/Repository/GingerSolutionRepository.cs:70]: Please check if we load RQM or Jira settings, same folder is being used and new one is not getting created

File: Ginger/GingerCoreCommon/WorkSpaceLib/Solution.cs [hunk 14]
```diff
@@ -78,6 +79,27 @@ public string Name
         [IsSerializedForLocalRepository]
         public ObservableList<RepositoryItemTag> Tags;
 
+
+        private AccessibilityConfiguration mDefaultExcludeRule;
```
Comment (Maheshkale447) [inline Ginger/GingerCoreCommon/WorkSpaceLib/Solution.cs:83]: Remove accessibility rules reference from solution.cs as it is not being used

**Review by Maheshkale447 — CHANGES_REQUESTED** [review]

File: Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs (hunk 9 above)
Comment (prashelke) [reply Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs:109]: Done

File: Ginger/GingerCoreCommon/External/Configurations/AccessibilityConfiguration.cs (hunk 11 above)
Comment (prashelke) [reply Ginger/GingerCoreCommon/External/Configurations/AccessibilityConfiguration.cs:41]: Done

File: Ginger/GingerCoreCommon/WorkSpaceLib/Solution.cs (hunk 14 above)
Comment (prashelke) [reply Ginger/GingerCoreCommon/WorkSpaceLib/Solution.cs:83]: Removed

File: Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs (hunk 8 above)
Comment (prashelke) [reply Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs:101]: Done

File: Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs (hunk 7 above)
Comment (prashelke) [reply Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs:51]: Done

File: Ginger/GingerCoreNET/ActionsLib/UI/Web/ActAccessibilityTesting.cs [hunk 15]
```diff
@@ -356,6 +372,13 @@
             All = 15
         }
 
+        public override eImageType Image { get { return eImageType.Accessibility; } }
+
+        private static string GetAccessiblityrules()
+        {
+            return EmbeddedResourceProvider.ReadEmbeddedFile("AccessiblityRules.json");
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/ActionsLib/UI/Web/ActAccessibilityTesting.cs:380]: done

File: Ginger/GingerCoreCommon/Repository/GingerSolutionRepository.cs (hunk 13 above)
Comment (prashelke) [reply Ginger/GingerCoreCommon/Repository/GingerSolutionRepository.cs:70]: Done

File: Ginger/GingerCoreNET/ActionsLib/UI/Web/ActAccessibilityTesting.cs [hunk 16]
```diff
@@ -303,11 +305,25 @@
...
+                {
+                    if (mAccessibilityConfiguration.DefaultExcludeRule != null && mAccessibilityConfiguration.DefaultExcludeRule.Select(x=>x.RuleID).Contains(ruleData.RuleID))
+                    {
+                        ruleData.Active = false;
+                    }
+                    else
+                    {
+                        ruleData.Active = true;
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/ActionsLib/UI/Web/ActAccessibilityTesting.cs]: done

File: Ginger/GingerCoreNET/GeneralLib/General.cs [hunk 17]
```diff
@@ -615,6 +616,20 @@ public static string CreateReportLinkPerFlow(string HtmlReportUrl, string Execut
...
+            {
+                AccessibilityConfiguration newAccessibilityConfiguration = new AccessibilityConfiguration() { Name = "Accessibility" };
+                WorkSpace.Instance.SolutionRepository.AddRepositoryItem(newAccessibilityConfiguration);
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/GeneralLib/General.cs]: done

File: Ginger/GingerCoreCommon/External/Configurations/AccessibilityConfiguration.cs [hunk 18]
```diff
@@ -29,231 +30,42 @@
...
             }
             set
             {
-                if (value != mOperationValueList)
+                if (mDefaultExcludeRule != value)
                 {
-                    mOperationValueList = value;
-                    OnPropertyChanged(nameof(OperationValueList));
+                    mDefaultExcludeRule = value;
+                    OnPropertyChanged(nameof(DefaultExcludeRule));
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreCommon/External/Configurations/AccessibilityConfiguration.cs]: done

File: Ginger/GingerCoreCommon/External/Configurations/AccessibilityConfiguration.cs [hunk 19]
```diff
@@ -29,231 +30,42 @@ namespace Ginger.Configurations
...
-        public Dictionary<string, object> Items
-        {
-            get
-            {
-                return _items;
-            }
-            set
-            {
-                _items = value;
+        public string Name { get { return mName; } set { if (mName != value) { mName = value; OnPropertyChanged(nameof(Name)); } } }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreCommon/External/Configurations/AccessibilityConfiguration.cs:35]: done

File: Ginger/Ginger/SolutionWindows/AccessiblityRulePage.xaml.cs [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+            xAccessiblityRulesGrid.InitViewItems();
+        }
+
+        private void LoadGridData()
+        {
+            ActAccessibilityTesting actAccessibilityTesting = new ActAccessibilityTesting();
+            List<AccessibilityRuleData> sortedList = actAccessibilityTesting.RulesItemsdata.OrderByDescending(data => !data.Active).ToList();
+            ObservableList<AccessibilityRuleData> accessibilityRuleDatas = [.. sortedList];
+            xAccessiblityRulesGrid.DataSourceList = accessibilityRuleDatas;
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/SolutionWindows/AccessiblityRulePage.xaml.cs]: done

File: Ginger/GingerCoreNETUnitTest/ActionsLib/UI/Web/AccessiblityTests.cs [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+        [ClassInitialize()]
+        public static void ClassInit(TestContext TC)
+        {
+            //// Initialize the WebDriver (Chrome in this case)
+            _driver = new ChromeDriver();
+            mact = new ActAccessibilityTesting();
+            mact.GetOrCreateInputParam(ActAccessibilityTesting.Fields.Target, ActAccessibilityTesting.eTarget.Page);
+            mact.SeverityList = GetSeverityList();
+            mact.StandardList = GetStandardTagslist();
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNETUnitTest/ActionsLib/UI/Web/AccessiblityTests.cs]: done

File: Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+{
+    /// <summary>
+    /// Interaction logic for AccessibilityRulePage.xaml
+    /// </summary>
+    public partial class AccessibilityRulePage : Page
+    {
+        Solution mSolution;
+        string AppName;
+        private AccessibilityConfiguration mAccessibilityConfiguration;
+        private static ActAccessibilityTesting actAccessibilityTesting;
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs]: done

File: Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+            xAccessibilityRulesGrid.SetAllColumnsDefaultView(view);
+            xAccessibilityRulesGrid.InitViewItems();
+        }
+
+        private void LoadGridData()
+        {
+            List<AccessibilityRuleData> sortedList = actAccessibilityTesting.RulesItemsdata.OrderByDescending(data => !data.Active).ToList();
+            ObservableList<AccessibilityRuleData> accessibilityRuleDatas = [.. sortedList];
+            xAccessibilityRulesGrid.DataSourceList = accessibilityRuleDatas;
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/SolutionWindows/AccessibilityRulePage.xaml.cs:103]: done

**Review by Maheshkale447 — DISMISSED** [review]

**Pull request merged**